pygsti/algorithms/germselection.py

```diff
@@ -1045,16 +1045,31 @@ def _super_op_for_perfect_twirl(wrt, eps, float_type=_np.cdouble):
     dim = wrt.shape[0]
     
     #The eigenvalues and eigenvectors of wrt can be complex valued, even for
-    #real-valued transfer matrices. Need to be careful here to start off using able
+    #real-valued transfer matrices. Need to be careful here to start off using a
     #complex data type. The actual projector onto the germs commutant appears to be strictly real valued though
     #(that makes sense because otherwise the projected derivative would become complex
     #So we should be able to cast it back to the specified float_type just before returning it.
     SuperOp = _np.zeros((dim**2, dim**2), dtype=_np.cdouble)
-
-    # Get spectrum and eigenvectors of wrt
-    wrtEvals, wrtEvecs = _np.linalg.eig(wrt)
-    wrtEvecsInv = _np.linalg.inv(wrtEvecs)
     
+    #Test the wrt matrix to see if it is normal. If so use the schur
+    #decomposition, otherwise use the general eigenvalue decomposition.
+    wrt_conj_transpose = wrt.conj().T
+    wrt_commutator = wrt@wrt_conj_transpose - wrt_conj_transpose@wrt
+    
+    # Get spectrum and eigenvectors of wrt
+    # May as well use the same eps here too.
+    if _np.linalg.norm(wrt_commutator) < eps:
```

Comment:
Two comments.

1. I think np.linalg.norm defaults to Frobenius. If that's the case then we should check that the norm is ``<= (dim**0.5) * tol`` for some value of ``tol``.

2. I'd like for ``tol`` to be some fractional power of the working precision's unit roundoff. I think a safe value for the power in this context is 0.75. For double precision arithmetic the roundoff is about 1e-16, so I'd suggest ``tol = (1e-16)**0.75 = 1e-12``. (When I was debugging this code I saw ``eps`` as large as 1e-6.)

Reply:
Thanks for the great suggestion, @rileyjmurray. I have just pushed a new commit implementing this change.